File: code/align.m

function  [mappedX,B,I,obj]= align(X, label,no_dims, k, eig_impl)
%LTSA Runs the local tangent space alignment algorithm
%
%   mappedX = ltsa(X, no_dims, k, eig_impl)
%
% The function runs the local tangent space alignment algorithm on dataset
% X, reducing the data to dimensionality d. The number of neighbors is
% specified by k.
%
%

% This file is part of the Matlab Toolbox for Dimensionality Reduction.
% The toolbox can be obtained from http://homepage.tudelft.nl/19j49
% You are free to use, change, or redistribute this code in any way you
% want for non-commercial purposes. However, it is appreciated if you 
% maintain the name of the original author.
%
% (C) Laurens van der Maaten, Delft University of Technology
for t = 1:length(unique(label))
Branch{t} = find(label==t);
end

    if ~exist('no_dims', 'var')
        no_dims = 2;
    end
    if ~exist('k', 'var')
        k = 12;
    end
    if ~exist('eig_impl', 'var')
        eig_impl = 'Matlab';
    end
 
    % Compute neighborhood indices
    if ischar(k)
        warning('Adaptive neighborhood selection often leads to problems in LTSA.');
    end
    disp('Find nearest neighbors...');
    n = size(X, 1);
    [D, ni] = myfind_nn(X, k);
%     [D, maxk,no_dim] = find_nn_adaptive(X);
    for i=1:n
%         I0{i} = ni(i,:);
        Ii = ni(i,:);%intersect(ni(i,:),Branch{label(i)});
%         Ii = find(D(i,:)~=0);
        Ii = Ii(Ii ~= 0);
        if label(i)==2
            Ii = setdiff(Ii,Branch{3});

        end
        if label(i)==3
            Ii = setdiff(Ii,Branch{2});
        end
%         if label(i)==2
%             Ii = setdiff(Ii,union(Branch{3},Branch{4}));
%         end
%         if label(i)==3
%             Ii = setdiff(Ii,union(Branch{2},Branch{4}));
%         end
%         if label(i)==4
%             Ii = setdiff(Ii,union(Branch{2},Branch{3}));
%         end
        I{i} = Ii;
    end


    
    % Compute local information matrix for all datapoints
    disp('Compute local information matrices for all datapoints...');
    Bi = cell(1, n); 
    for i=1:n
        % Compute correlation matrix W
        Ii = I{i};
        kt = numel(Ii);
        Xi = X(Ii,:) - repmat(mean(X(Ii,:), 1), [kt 1]);
        W = Xi * Xi'; 
        W = (W + W') / 2;
        
        % Compute local information by computing d largest eigenvectors of W
        [Vi, Si] = schur(W);
        [s, Ji] = sort(-diag(Si));
		if length(Ji) < no_dims
			no_dims = length(Ji);
			warning(['Target dimensionality reduced to ' num2str(no_dims) '...']);
		end
        Vi = Vi(:,Ji(1:no_dims)); 
        
        if 1
%         % Store eigenvectors in G (Vi is the space with the maximum variance, i.e. a good approximation of the tangent space at point Xi)
% 		% The constant 1/sqrt(kt) serves as a centering matrix
		Gi = double([repmat(1 / sqrt(kt), [kt 1]) Vi]);
%         % Compute Bi = I - Gi * Gi'
		Bi{i} = eye(kt) - Gi * Gi';  
        else
            Pi = eye(kt) - Vi*Vi';
            Wi = Pi - repmat(mean(Pi, 1), [kt 1]);
            Bi{i} = Wi*Wi';
        end
    end
    
    % Construct sparse matrix B (= alignment matrix)
    disp('Construct alignment matrix...');
    B = speye(n);
    for i=1:n
%         Ii = intersect(ni(i,:),Branch{label(i)});;
%         Ii = Ii(Ii ~= 0);
        Ii = I{i};
        B(Ii, Ii) = B(Ii, Ii) + Bi{i};							% sum Bi over all points
		B(i, i) = B(i, i) - 1;
    end
	B = (B + B') / 2;											% make sure B is symmetric
	B = full(B);
    
	% For sparse datasets, we might end up with NaNs in M. We just set them to zero for now...
	B(isnan(B)) = 0;
	B(isinf(B)) = 0;
    
    % Perform eigenanalysis of matrix B
	disp('Perform eigenanalysis...');
    tol = 0;
	if strcmp(eig_impl, 'JDQR')
        options.Disp = 0;
        options.LSolver = 'bicgstab';
        [mappedX, D] = jdqr(B, no_dims + 1, tol, options);      % only need bottom (no_dims + 1) eigenvectors
    else
        options.disp = 0;
        options.isreal = 1;
        options.issym = 1;
        [mappedX, D] = eigs(B, no_dims + 1, tol, options);      % only need bottom (no_dims + 1) eigenvectors
    end

    % Sort eigenvalues and eigenvectors
    [D, ind] = sort(diag(D), 'ascend');
    mappedX = mappedX(:,ind);

    % Final embedding coordinates
	if size(mappedX, 2) < no_dims + 1, no_dims = size(mappedX, 2) - 1; end
    mappedX = mappedX(:,2:no_dims + 1);
    
    for i = 1:n
        obj(i) = norm(Bi{i}*mappedX(I{i},:),'fro')^2;
    end
    obj = obj';
%     sumobj = sum(obj);
    

File: code/align.R
align <- function(X, label, root, parameters){
  n = nrow(X)
  d = 1
  
  labelset = unique(label)
  labelset = sort(labelset)
  Branch<-array(NA, dim = c(1, nrow(label), length(unique(label))))
  for (t in 1:length(labelset)){
    st=which(label==t)
    if (length(st)>0){
      Branch[1,1:length(st) ,labelset[t]]=st
    }else{
      Branch[1,,labelset[t]]=NA
    }
  }
  
  rootlabel = label[root,]
  nonrootlabel = setdiff(labelset,rootlabel)
  RootBranch = Branch[,,rootlabel]
  RootBranch = na.omit(RootBranch)
  
  
  
  
  D0 = 999*matrix(1,nrow = n,ncol = n)
  D0[RootBranch,] = 0
  D0[,RootBranch] = 0
  
  for (t in 1:length(nonrootlabel)){
    a = nonrootlabel[t]
    Brancha = Branch[,,a]
    Brancha=na.omit(Brancha)
    D0[Brancha,Brancha] = 0
    D0[Brancha,RootBranch] = 0
    D0[RootBranch,Brancha] = 0
  }
  
  D = DISTMAT(X)
  D = D + D0
  
  
  X = t(X)
  
  J <- matrix(nr=nrow(D),nc=ncol(D))
  for (i in 1:nrow(D)){
    J[i,] <- order(D[i,])
  }
  
  
  
  skip = 1
  kmin = parameters[1]
  Kmax = parameters[2]
  Kmax = min(Kmax,n)
  eta = parameters[3]
  
  Is<-array(list(), n)
  
  
  for (i in 1:n){
    # Contraction
    ki = Kmax
    ratio = rep(0,Kmax)
    while (ki>=kmin){
      Ii = J[i,1:ki]
      Xi = X[,Ii]
      xmean = rowMeans(Xi)
      Xi =  Xi - matrix(rep(xmean, ki), ncol = ki)
      si = svd(Xi)
      si = si[["d"]]
      ratio[ki] = norm(si[(d+1):length(si)],'2')/norm(si[1:d],'2')
      if (ratio[ki]<eta){
        break
      }else{
        ki = ki-skip
      }
    }
    
    if (ki<kmin){
      j = which.min((ratio[kmin:Kmax]))
      ki =  ki+skip-1+j
      Ii = J[i,1:ki]
      Xi = X[,Ii]
      xmean = rowMeans(Xi)
      Xi =  Xi - matrix(rep(xmean, ki), ncol = ki)
    }
    Qi = svd(Xi)
    Qi = Qi[["u"]]
    Qi = Qi[,1:d]
    
    # Expansion
    if(ki+1<=Kmax){
      for (j in (ki+1):Kmax){
        xj = X[,J[i,j]]-xmean
        thetaj = t(Qi) %*% xj
        if (norm(xj-Qi %*% thetaj)< eta *norm(thetaj)){
          Ii = c(Ii,J[i,j])
        }
      }
    }
    Is[[i]]=Ii
    
  }
  
  
  X = t(X)
  # Compute local information matrix for all datapoints
  print('Compute local information matrices for all datapoints...')
  Bi = array(list(), n)
  K = numeric(n)
  for (i in 1:n) {
    K[i] = length(Is[[i]])
  }
  
  for (i in 1:n){
    # Compute correlation matrix W
    Ii = Is[[i]]
    kt = K[i]
    Xi = X[Ii,] - matrix(rep(colMeans(X[Ii, ]), kt), nrow = kt, byrow = TRUE)
    W = Xi %*% t(Xi)
    W = (W + t(W))/2
    
    # Compute local information by computing d largest eigenvectors of W
    tmp = eigs_sym(W, d, which = "LM")
    Vi = tmp$vectors[,1:d]
    
    Pi = diag(rep(1,kt)) - Vi %*% t(Vi)
    Wi = Pi - matrix(rep(colMeans(Pi), kt), nrow = kt, byrow = TRUE)
    Bi[[i]] = Wi %*% t(Wi)
    
  }
  
  
  # Construct sparse matrix B (= alignment matrix)
  print("Construct alignment matrix...")
  B = diag(rep(1, n))
  for (i in 1:n){
    Ii = Is[[i]]
    B[Ii,Ii] = B[Ii,Ii] + Bi[[i]]  # sum Bi over all points
    B[i,i] = B[i,i] - 1
    
  }
  B = (B + t(B))/2                                          # make sure B is symmetric
  
  
  # For sparse datasets, we might end up with NaNs in M. We just set them to zero for now...
  B[B==-Inf]<-0
  B[is.na(B)]<-0
  
  B <- as(B,'dgCMatrix')
  
  # Perform eigenanalysis of matrix B
  print("Perform eigenanalysis...")
  
  eigB = eigs_sym(B, d+10, which = "SM")
  ind = order(eigB$values, decreasing = FALSE)
  
  # Final embedding coordinates
  D = eigB$values[ind[2:(d + 1)]]
  mappedX = eigB$vectors[, ind[2:(d + 1)]]
  T = mappedX
  
  
  
  picknonrootbranch = Branch[,,nonrootlabel[1]]
  picknonrootbranch = na.omit(picknonrootbranch)
  nonroot = picknonrootbranch[1]
  
  a = min(T)
  b = max(T)
  T = (T-a)/(b-a)
  
  id = which(T<T[root])
  T[id] = 2*T[root]-T[id]
  
  
  pseudo = T
  return(pseudo)
  
}
